decoder/context.go

```diff
@@ -16,8 +21,79 @@ type DecoderContext struct {
 	// CodeLenses represents a slice of executable lenses
 	// which will be executed in the exact order they're declared
 	CodeLenses []lang.CodeLensFunc
+
+	CompletionHooks        CompletionFuncMap
+	CompletionResolveHooks CompletionResolveFuncMap
 }
 
 func (d *Decoder) SetContext(ctx DecoderContext) {
 	d.ctx = ctx
 }
+
+type CompletionFunc func(ctx context.Context, value cty.Value) ([]Candidate, error)
+type CompletionFuncMap map[string]CompletionFunc
+
+type CompletionResolveFuncMap map[string]CompletionResolveFunc
+type CompletionResolveFunc func(ctx context.Context, unresolvedCandidate UnresolvedCandidate) (*ResolvedCandidate, error)
+
+type Candidate struct {
+	// Label represents a human-readable name of the candidate
+	// if one exists (otherwise Value is used)
+	Label string
+
+	// Detail represents a human-readable string with additional
+	// information about this candidate, like symbol information.
+	Detail string
+
+	Kind lang.CandidateKind
+
+	// Description represents human-readable description
+	// of the candidate
+	Description lang.MarkupContent
+
+	// IsDeprecated indicates whether the candidate is deprecated
+	IsDeprecated bool
+
+	RawInsertText string
```

Comment:
Do you mind adding a comment to briefly explain what it's for and how we treat the value in terms of quoting - i.e. that we expect it to be already quoted, if necessary? Just so that consumers know and don't have to find out via trial & error or by studying the implementation 😅